### PR TITLE
[fix][broker] Fix skip message API when hole messages exists

### LINKED ISSUE #20262: [Bug] When using the admin.topics().skipMessages API, the actual number of skipped messages may be less than the expected number when the value of skipNumber is greater than 8.
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

latest version

### Minimal reproduce step

```java
    @Test(dataProvider = "topicName")
    public void testSkipMessages(String topicName) throws Exception {
        final String subName = topicName;
        assertEquals(admin.topics().getList("prop-xyz/ns1"), new ArrayList<>());

        final String persistentTopicName = "persistent://prop-xyz/ns1/" + topicName;
        // Force to create a topic
        publishMessagesOnPersistentTopic("persistent://prop-xyz/ns1/" + topicName, 0);
        assertEquals(admin.topics().getList("prop-xyz/ns1"),
                List.of("persistent://prop-xyz/ns1/" + topicName));

        // create consumer and subscription
        @Cleanup
        PulsarClient client = PulsarClient.builder()
                .serviceUrl(pulsar.getWebServiceAddress())
                .statsInterval(0, TimeUnit.SECONDS)
                .build();
        AtomicInteger total = new AtomicInteger();
        Consumer<byte[]> consumer = client.newConsumer().topic(persistentTopicName)
                .messageListener(new MessageListener<byte[]>() {
                    @SneakyThrows
                    @Override
                    public void received(Consumer<byte[]> consumer, Message<byte[]> msg) {
                        if (total.get() %2 !=0){
                            consumer.acknowledge(msg);
                        }
                        total.incrementAndGet();
                    }
                })
                .subscriptionName(subName)
                .subscriptionType(SubscriptionType.Exclusive).subscribe();

        assertEquals(admin.topics().getSubscriptions(persistentTopicName), List.of(subName));

        publishMessagesOnPersistentTopic("persistent://prop-xyz/ns1/" + topicName, 100);
        TimeUnit.SECONDS.sleep(3);
        TopicStats topicStats = admin.topics().getStats(persistentTopicName);
        long msgBacklog = topicStats.getSubscriptions().get(subName).getMsgBacklog();
        log.info("back={}",msgBacklog);
        int skipNumber = 20;
        admin.topics().skipMessages(persistentTopicName, subName, skipNumber);
        topicStats = admin.topics().getStats(persistentTopicName);
        assertEquals(topicStats.getSubscriptions().get(subName).getMsgBacklog(), msgBacklog - skipNumber);
    }
```

**When I artificially created 50 hollow messages** and skipped 20 messages, only 14 messages were actually skipped.

### What did you expect to see?

The expected number of skipped messages is equal to the actual number of skipped messages.

### What did you see instead?

The actual number of skips is less than the expected number of skips.
image

### Anything else?

https://github.com/apache/pulsar/blob/e956db729f5098ff319237fd4220dbfb234c1b18/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java#L1772-L1785
image

The reason for this issue is that the `recycle()` function reuses objects, causing the object referenced by `r` to change during runtime. When the loop count is greater than 8, the else block is entered, leading to an incorrect calculation of the message count.

I think there are two methods to fix this issue.

The first method is to perform a copy before assignment, similar to the following:

```java
        public void setStartPosition(PositionImpl startPosition) {
            PositionImpl cp = new PositionImpl(startPosition.ledgerId, startPosition.entryId);
            this.startPosition = cp;
        }

        public void setEndPosition(PositionImpl endPosition) {
            PositionImpl cp = new PositionImpl(endPosition.ledgerId, endPosition.entryId);
            this.endPosition = cp;
        }
	
	// state.endPosition = r.lowerEndpoint();
	state.setEndPosition(r.lowerEndpoint());

	// state.startPosition = r.upperEndpoint();
	state.setStartPosition(r.upperEndpoint());
```

The second method is to remove `recyclePositionRangeConverter`.

image

any other suggestions for a better solution?

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by crossoverJie**
Fixes #20262

### Motivation

When there are hole messages, there is a bug in the API of skipping the message, please refer to the [issue](https://github.com/apache/pulsar/issues/20262) for the specific reason
### Modifications
Remove `recyclePositionRangeConverter`,  revert to normal loop.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: (https://github.com/crossoverJie/pulsar/pull/8)

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 1]
```diff
@@ -1757,32 +1757,26 @@ long getNumIndividualDeletedEntriesToSkip(long numEntries) {
...
+                    Range<PositionImpl> range = Range.openClosed(state.startPosition, state.endPosition);
+                    long entries = ledger.getNumberOfEntries(range);
+                    if (state.totalEntriesToSkip + entries >= numEntries) {
+                        // do not process further
+                        return false;
                     }
-                    return true;
-                } finally {
-                    if (r.lowerEndpoint() instanceof PositionImplRecyclable) {
-                        ((PositionImplRecyclable) r.lowerEndpoint()).recycle();
```
Comment (liangyepianzhou) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:1781]: We can just recycle r.lowerEndpoint() here.

Comment (crossoverJie) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:1781]: Good suggestion

Comment (crossoverJie) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:1781]: Good suggestion

**Review by liangyepianzhou — APPROVED** [review]
LGTM

**Review by liangyepianzhou — APPROVED** [review]
LGTM

**Comment by liangyepianzhou** [comment]
@poorbarcode @Technoboy- Please help take a look, thanks. 😸

**Review by Technoboy- — APPROVED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 2]
```diff
@@ -1779,7 +1779,6 @@ long getNumIndividualDeletedEntriesToSkip(long numEntries) {
                 } finally {
                     if (r.lowerEndpoint() instanceof PositionImplRecyclable) {
                         ((PositionImplRecyclable) r.lowerEndpoint()).recycle();
```
[the remark this answers is not in the document]
Comment (liangyepianzhou) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:1781]: We recycle an object after it is no longer used to save GC pause.
But the upperEndpoint is still used after being recycled, which makes this issue.
The lowerEndpoint is no longer used after being recycled, so we can recycle it.
```java
state.startPosition = r.upperEndpoint();
```

**Comment by michaeljmarshall** [comment]
As discussed on the mailing list https://lists.apache.org/thread/w4jzk27qhtosgsz7l9bmhf1t7o9mxjhp, there is no plan to release 2.9.6, so I am going to remove the release/2.9.6 label

**Pull request merged**